File: leetcode/sliding-window-maximum.py
class Solution {
public:
    struct Stack{
        stack<int> s, sMax;
        bool empty() const {
            return s.empty();
        }
        void push(int i) {
            sMax.push(std::max(max(), i));
            s.push(i);
        }
        int pop() {
            sMax.pop();
            auto x = s.top();
            s.pop();
            return x;
        }

        int max() const {
            if (empty()) return INT_MIN;
            return sMax.top();
        }
    };
    
    Stack sStart, sEnd;

    void add(int i) {
        sEnd.push(i);
    }

    void remove() {
        if (sStart.empty()) {
            while (!sEnd.empty()) {
                sStart.push(sEnd.pop());
            }
        }
        sStart.pop();
    }

    int maxElem() {
        return max(sStart.max(), sEnd.max());
    }

    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if (nums.size() < k) return {};
        for (int i = 0; i < k - 1; ++i) {
            add(nums[i]);
        }
        vector<int> res;
        for (int r = k - 1; r < nums.size(); ++r) {
            add(nums[r]);
            res.push_back(maxElem());
            remove();
        }
        return res;
    }
};

File: leetcode/sliding-window-maximum.cpp
#include <vector>
#include <stack>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
    struct Stack {
        stack<int> s, sMax;

        void push(int a) {
            s.push(a);
            int prevMax = sMax.empty() ? INT_MIN : sMax.top();
            sMax.push(max(a, prevMax));
        }

        void pop() {
            s.pop();
            sMax.pop();
        }

        bool empty() const {
            return s.empty();
        }

        int maxElem() const {
            return sMax.empty() ? INT_MIN : sMax.top();
        }

        int top() const {
            return s.top();
        }
    };

    Stack sStart, sEnd;

    void add(int a) {
        sEnd.push(a);
    }

    void remove() {
        if (sStart.empty()) {
            while (!sEnd.empty()) {
                sStart.push(sEnd.top());
                sEnd.pop();
            }
        }
        sStart.pop();
    }

    int maxElemBoth() {
        return max(sStart.maxElem(), sEnd.maxElem());
    }

public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        if (k <= 0) return res;

        for (int i = 0; i < k - 1; ++i) add(nums[i]);

        for (int i = k - 1; i < (int)nums.size(); ++i) {
            add(nums[i]);
            res.push_back(maxElemBoth());
            remove();
        }
        return res;
    }
};